src/core/thread/mle.cpp

```diff
@@ -1879,6 +1986,213 @@ ThreadError Mle::HandleChildUpdateResponse(const Message &aMessage, const Ip6::M
+
+    SuccessOrExit(error = SendMessage(*message, aDestination));
+
+    otLogInfoMle("Sent discovery response\n");
+
+exit:
+
+    if (error != kThreadError_None && message != NULL)
+    {
+        Message::Free(*message);
+    }
+
+    return error;
+}
+
+ThreadError Mle::HandleDiscoveryResponse(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    ThreadError error = kThreadError_None;
+    const ThreadMessageInfo *threadMessageInfo = reinterpret_cast<const ThreadMessageInfo *>(aMessageInfo.mLinkInfo);
+    Tlv tlv;
+    MeshCoP::Tlv meshcopTlv;
+    MeshCoP::DiscoveryResponseTlv discoveryResponse;
+    MeshCoP::ExtendedPanIdTlv extPanId;
+    MeshCoP::NetworkNameTlv networkName;
+    otActiveScanResult result;
+    uint16_t offset;
+    uint16_t end;
+    char networkNameBuf[OT_NETWORK_NAME_SIZE];
+
+    otLogInfoMle("Handle discovery response\n");
+
+    offset = aMessage.GetOffset();
+    end = aMessage.GetLength();
+
+    // find MLE Discovery TLV
+    while (offset < end)
+    {
+        aMessage.Read(offset, sizeof(tlv), &tlv);
+
+        if (tlv.GetType() == Tlv::kDiscovery)
+        {
+            break;
+        }
+
+        offset += sizeof(tlv) + tlv.GetLength();
+    }
+
+    VerifyOrExit(offset < end, error = kThreadError_Parse);
+
+    offset += sizeof(tlv);
+    end = offset + sizeof(tlv) + tlv.GetLength();
+
+    memset(&result, 0, sizeof(result));
+    result.mPanId = threadMessageInfo->mPanId;
+    result.mChannel = threadMessageInfo->mChannel;
+    result.mRssi = threadMessageInfo->mRss;
+    result.mLqi = threadMessageInfo->mLqi;
+    static_cast<Mac::ExtAddress *>(&result.mExtAddress)->Set(aMessageInfo.GetPeerAddr());
+
+    // process MeshCoP TLVs
+    while (offset < end)
+    {
+        aMessage.Read(offset, sizeof(meshcopTlv), &meshcopTlv);
+
+        switch (meshcopTlv.GetType())
+        {
+        case MeshCoP::Tlv::kDiscoveryResponse:
+            aMessage.Read(offset, sizeof(discoveryResponse), &discoveryResponse);
+            VerifyOrExit(discoveryResponse.IsValid(), error = kThreadError_Parse);
+            result.mVersion = discoveryResponse.GetVersion();
+            result.mIsNative = discoveryResponse.IsNativeCommissioner();
+            break;
+
+        case MeshCoP::Tlv::kExtendedPanId:
+            aMessage.Read(offset, sizeof(extPanId), &extPanId);
+            VerifyOrExit(extPanId.IsValid(), error = kThreadError_Parse);
+            result.mExtPanId = extPanId.GetExtendedPanId();
+            break;
+
+        case MeshCoP::Tlv::kNetworkName:
+            aMessage.Read(offset, sizeof(networkName), &networkName);
+            VerifyOrExit(networkName.IsValid(), error = kThreadError_Parse);
+            memcpy(networkNameBuf, networkName.GetNetworkName(), networkName.GetLength());
+            memset(networkNameBuf + networkName.GetLength(), 0, sizeof(networkNameBuf) - networkName.GetLength());
+            result.mNetworkName = networkNameBuf;
+            break;
+
+        default:
```

Comment:
Do we need to add steering data in Discovery, though it is related with commissioning? The node would not be able to join if steering data is zero or not included.
If it would be implemented in commissioning, please ignore this comment.